mergekit/card.py

```diff
@@ -0,0 +1,182 @@
+# Copyright (C) 2024 Charles O. Goddard
+#
+# This software is free software: you can redistribute it and/or
+# modify it under the terms of the GNU Lesser General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This software is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program. If not, see http://www.gnu.org/licenses/.
+
+import os
+from collections.abc import Iterable
+from typing import Any, List, Optional, Sequence
+
+import huggingface_hub
+import yaml
+from yaml.nodes import SequenceNode as SequenceNode
+
+from mergekit.config import MergeConfiguration, ModelReference
+
+CARD_TEMPLATE = """---
+{metadata}
+---
+# {name}
+
+This is a merge of pre-trained language models created using [mergekit](https://github.com/cg123/mergekit).
+
+## Merge Details
+### Merge Method
+
+This model was merged using the {merge_method} merge method{base_text}.
+
+### Models Merged
+
+The following models were included in the merge:
+{model_list}
+
+### Configuration
+
+The following YAML configuration was used to produce this model:
+
+```yaml
+{config_yaml}
+```
+"""
+
+
+class ConfigYamlDumper(yaml.Dumper):
+    """Custom YAML dumper to format lists of numbers in flow style."""
+
+    def represent_list(self, data: Iterable[Any]) -> SequenceNode:
+        flow_style = all(isinstance(e, (int, float)) for e in data)
+        return self.represent_sequence(
+            "tag:yaml.org,2002:seq", data, flow_style=flow_style
+        )
+
+
+ConfigYamlDumper.add_representer(list, ConfigYamlDumper.represent_list)
+
+
+def is_hf(path: str) -> bool:
+    """
+    Determines if the given path is a Hugging Face model repository.
+
+    Args:
+        path: A string path to check.
+    """
+    if path[0] in "/~" or path.count("/") > 1:
+        return False  # definitely a local path
+    if not os.path.exists(path):
+        return True  # If path doesn't exist locally, it must be a HF repo
+    return huggingface_hub.repo_exists(path, repo_type="model")
```

Comment:
Will almost all models included in merges already be publicly shared? I doubt this will often be an issue, but this method will return `True` for private models to which the user has access if a valid Hub token is found. For now, I think it's fine to leave this, but you could add a `token=False` argument if you want to ensure only public repo returns True for this.

Reply:
Good point. Since reading your comment I've gone back and forth a few times on whether private repos should be listed in `base_model`. I'll go with `token=False` to avoid the appearance of "broken" links for now.